tests/emails_s3_tests.rs

```diff
@@ -0,0 +1,317 @@
+#[allow(dead_code)]
+mod helpers;
+
+use std::collections::HashMap;
+use std::env;
+use std::io::Write;
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+
+use tokio::runtime::{Builder, Runtime};
+
+use chrono::prelude::*;
+use dipstick::{Input, Prefixed, Statsd};
+use serde_json::json;
+use serial_test::serial;
+use tokio_util::sync::CancellationToken;
+use uuid::Uuid;
+
+use kafka_delta_ingest::{instrumentation::StatsHandler, KafkaJsonToDelta, Options};
+use rusoto_core::Region;
+use rusoto_s3::{CopyObjectRequest, S3};
+use tokio::task::JoinHandle;
+
+const TEST_S3_ENDPOINT: &str = "http://localhost:4566";
+const TEST_S3_BUCKET: &str = "tests";
+const TEST_APP_ID: &str = "emails_test";
+const TEST_BROKER: &str = "0.0.0.0:9092";
+const TEST_CONSUMER_GROUP_ID: &str = "kafka_delta_ingest_emails";
+const TEST_PARTITIONS: i32 = 4;
+const TEST_TOTAL_MESSAGES: i32 = 100;
+
+const WORKER_1: &str = "WORKER-1";
+const WORKER_2: &str = "WORKER-2";
+
+#[tokio::test(flavor = "multi_thread")]
+#[serial]
+async fn when_both_workers_started_simultaneously() {
+    run_emails_s3_tests(false).await;
+}
+
+#[tokio::test(flavor = "multi_thread")]
+#[serial]
+async fn when_rebalance_happens() {
+    run_emails_s3_tests(true).await;
+}
+
+struct TestScope {
+    pub topic: String,
+    pub table: String,
+    pub workers_token: Arc<CancellationToken>,
+    pub runtime: HashMap<&'static str, Runtime>,
+}
+
+async fn run_emails_s3_tests(initiate_rebalance: bool) {
+    init_logger();
+    let scope = TestScope::new().await;
+
+    let w1 = scope.create_and_start(WORKER_1).await;
+
+    // in order to initiate rebalance we first send messages,
+    // ensure that worker 1 consumes  some of them and then crate worker 2,
+    // otherwise, to proceed without rebalance the two workers has to be created simultaneously
+    let w2 =
+        if initiate_rebalance {
+            scope.send_messages(TEST_TOTAL_MESSAGES).await;
+            thread::sleep(Duration::from_secs(2));
+            scope.create_and_start(WORKER_2).await
+        } else {
+            let w = scope.create_and_start(WORKER_2).await;
+            thread::sleep(Duration::from_secs(5));
+            scope.send_messages(TEST_TOTAL_MESSAGES).await;
+            w
+        };
```

Comment:
This is the most effortless solution to _track_ rebalance I've came up with so far. So the idea is to start both workers and ensure that rebalance happens. However how to achieve that?
1. Try to log into file and grep logs?
2. Expose some "listener" api into source code to subscribe onto those events. Is it even a good practice? (hmm, however I'll look into that and find out whether it's possible to use dependency injection somehow)
3. any other ideas?

Reply:
I think 1. is an OK short term solution. We could also capture process stdout in tests and search for content in memory.

I am thinking we would need to emit metrics for rebalance events for monitoring purpose in production, so we could also send that metric to a dummy statsd process to capture the event. Or we can make the metric or event sending functionality pluggable and provide a dummy sender in tests.

Reply:
Created a ticket for https://github.com/delta-io/kafka-delta-ingest/issues/26